### ISSUE TITLE
use python3

### ISSUE THREAD

**Issue by KrisThielemans** [SyneRBI/SyneRBI_VM#122]
Python2 becomes more and more of a problem. We now fail to install spyder, see https://github.com/CCPPETMR/CCPPETMR_VM/issues/121.

Sadly, on Ubuntu 18.04, default `python` is still python2.7. I guess we can easily tell the SuperBuild to use python3. But what about `jupyter`? Will it just work when we use
```
python3 -m pip install jupyter
```

Possibly easiest would be to do
```
sudo update-alternatives --install /usr/bin/python python /usr/bin/python3.6 1
sudo update-alternatives  --set python /usr/bin/python3.6
```
before doing anything python-like at all.

@casperdcl @bathomas what do you think?

**Comment by KrisThielemans** [comment]
A complication might be that the SIRF-exercises notebooks had to include python2 kernel name, even though some of them have python3, e.g.
```
"metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
```
in https://raw.githubusercontent.com/CCPPETMR/SIRF-Exercises/master/notebooks/Reg/sirf_registration.ipynb

**Comment by casperdcl** [comment]
`update-alternatives` for a system library is risky; it may break something else. `jupyter` will "just work" (with a warning about selecting a new kernel in case the name changes)

**Comment by casperdcl** [comment]
also I presume ubuntu 20.04LTS will default to py3.

**Comment by Ede1994** [comment]
So I tested with a jupyter notebook based on python2 and the following code was issued by the terminal:
```
[IPKernelApp] ERROR | Failed to create history session in /home/sirfuser/.ipython/profile_default/history.sqlite. History will not be saved.
Traceback (most recent call last):
  File &quot;/usr/local/lib/python2.7/dist-packages/IPython/core/history.py&quot;, line 549, in __init__
    self.new_session()
  File &quot;&lt;/usr/local/lib/python2.7/dist-packages/decorator.pyc:decorator-gen-22&gt;&quot;, line 2, in new_session
  File &quot;/usr/local/lib/python2.7/dist-packages/IPython/core/history.py&quot;, line 60, in needs_sqlite
    return f(self, *a, **kw)
  File &quot;/usr/local/lib/python2.7/dist-packages/IPython/core/history.py&quot;, line 575, in new_session
    NULL, &quot;&quot;) &quot;&quot;&quot;, (datetime.datetime.now(),))
OperationalError: database is locked
```

**Comment by KrisThielemans** [comment]
This seems unrelated. Best to try again after a reboot of the VM.

**Comment by KrisThielemans** [comment]
@casperdcl you say
>jupyter in python3 wll just work

After `python3 -m pip install jupyter`, I presume that will override an existing `python2` install, and that the new one wlil have `python3` as default kernel? (A volunteer for trying this?)

**Comment by KrisThielemans** [comment]
The other thing though is to make sure that `CMake` picks up `python3` in the build. That might need a change in `UPDATE.sh` etc.

**Comment by KrisThielemans** [comment]
We're [using CMake 3.13](https://github.com/CCPPETMR/CCPPETMR_VM/blob/0e2e1161e691f38cb7889907aa1dbc6ecebda095/scripts/INSTALL_CMake.sh#L17). We should probably by using `FindPython` now, which would then default to using `python3` apparently.

See https://github.com/CCPPETMR/SIRF-SuperBuild/issues/313

**Comment by casperdcl** [comment]
> After python3 -m pip install jupyter, I presume that will override an existing python2 install, and that the new one wlil have python3 as default kernel?

Assuming

1) there's also still a py2 jupyter installed
2) The launch command is just `jupyter`
3) the launch is not within a conda/venv

Then it would depend on the order in PATH whether jupyter `py2` or `3` took precedence. Otherwise there would be no conflict.

For our purposes (vm/docker) we can definitely fully control this (incl. e.g. setting a default py2 kernel for py3 jupyter)

**Comment by KrisThielemans** [comment]
some other things to think about when upgrading:
- nbstripout is pinned
- spyder is pinned

**Comment by paskino** [comment]
I've removed the pin to [`nbstripout`](https://github.com/SyneRBI/SyneRBI_VM/blob/python3_build/scripts/UPDATE.sh#L143) and [`spyder` ](https://github.com/SyneRBI/SyneRBI_VM/blob/python3_build/scripts/INSTALL_python_packages.sh#L70)with python3, which you run with `spyder3`.

**Comment by KrisThielemans** [comment]
please create the PR.

**Comment by KrisThielemans** [comment]
closed with #170
